pallets/foreign-investments/src/lib.rs

```diff
@@ -43,43 +43,137 @@
+use cfg_traits::{investments::TrancheCurrency, PoolInspect};
+use cfg_types::investments::{CollectedAmount, Swap};
+use frame_support::{dispatch::DispatchResult, ensure};
 pub use pallet::*;
-
-pub mod errors;
-pub mod hooks;
-pub mod impls;
-pub mod types;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{traits::EnsureSub, ArithmeticError, DispatchError};
 
 #[cfg(test)]
 mod mock;
 
 #[cfg(test)]
 mod tests;
 
+/// Hold the base information of a foreign investment/redemption
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct BaseInfo<T: Config> {
+	foreign_currency: T::CurrencyId,
+	pool_currency: T::CurrencyId,
+	collected: CollectedAmount<T::Balance>,
+}
+
+impl<T: Config> BaseInfo<T> {
+	fn new(
+		investment_id: T::InvestmentId,
+		foreign_currency: T::CurrencyId,
+	) -> Result<Self, DispatchError> {
+		Ok(Self {
+			foreign_currency,
+			pool_currency: T::PoolInspect::currency_for(investment_id.of_pool())
+				.ok_or(Error::<T>::PoolNotFound)?,
+			collected: CollectedAmount::default(),
+		})
+	}
+
+	fn ensure_same_foreign(&self, foreign_currency: T::CurrencyId) -> DispatchResult {
+		ensure!(
+			self.foreign_currency == foreign_currency,
+			Error::<T>::MismatchedForeignCurrency
+		);
+
+		Ok(())
+	}
+}
+
+/// Hold the information of a foreign investment
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct InvestmentInfo<T: Config> {
+	base: BaseInfo<T>,
+	increase_swap_id: Option<T::SwapId>,
+	decrease_swap_id: Option<T::SwapId>,
+	total_pool_amount: T::Balance,
+}
+
+impl<T: Config> InvestmentInfo<T> {
+	fn new(
+		investment_id: T::InvestmentId,
+		foreign_currency: T::CurrencyId,
+	) -> Result<Self, DispatchError> {
+		Ok(Self {
+			base: BaseInfo::new(investment_id, foreign_currency)?,
+			total_pool_amount: T::Balance::default(),
+			increase_swap_id: None,
+			decrease_swap_id: None,
+		})
+	}
+
+	fn remaining_pool_amount(&self) -> Result<T::Balance, ArithmeticError> {
+		self.total_pool_amount
+			.ensure_sub(self.base.collected.amount_payment)
+	}
+}
+
+/// Hold the information of an foreign redemption
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct RedemptionInfo<T: Config> {
+	base: BaseInfo<T>,
+	swap_id: Option<T::SwapId>,
+	total_tranche_tokens: T::Balance,
+	swapped_amount: T::Balance,
+	tranche_tokens_sent: T::Balance,
+}
+
+impl<T: Config> RedemptionInfo<T> {
+	fn new(
+		investment_id: T::InvestmentId,
+		foreign_currency: T::CurrencyId,
+	) -> Result<Self, DispatchError> {
+		Ok(Self {
+			base: BaseInfo::new(investment_id, foreign_currency)?,
+			swap_id: None,
+			total_tranche_tokens: T::Balance::default(),
+			swapped_amount: T::Balance::default(),
+			tranche_tokens_sent: T::Balance::default(),
```

Comment:
I am having hard times to follow this whole new syntax, probably because it is a bit mixed here and there. Of course, I understand this is all not set in stone and thus not perfect. What does `sent` here refer to? The amount of tranche tokens which were collected during epoch fulfillment?

Reply:
Sorry, the naming for these variables may not be the best. I am open to suggestions. And I should add docs to them, too.

`tranche_tokens_sent` is the number of tranche tokens already notified by an `ExecutedForeignCollected.` I need to keep this information so I can set the correct value for the next `ExecutedForeignCollected` event, it is, the unnotified amount.

Reply:
Thanks for the explanation! Then I propose to call this `redeemed_` or `notified_tranche_tokens`.

Reply:
I have managed to eliminate that variable in this simplification: [80f5fa9](https://github.com/centrifuge/centrifuge-chain/pull/1698/commits/80f5fa91ecc4788083b7f213ec182988aa83560c)